File: src/test/java/com/springdebugger/convention/ConventionCatalogTest.java
package com.springdebugger.convention;

import org.junit.jupiter.api.Test;

import java.util.HashSet;
import java.util.Set;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Validates conventions.yaml: it loads, ids are unique, every checkType resolves to a registered
 * implementation, and the seed Javadoc rule is shaped as documented. No fixed rule count is asserted
 * so adding rules never breaks this test.
 */
class ConventionCatalogTest {

    private final ConventionCatalog catalog = ConventionCatalog.load();

    @Test
    void catalogLoadsAtLeastOneRule() {
        assertThat(catalog.all()).isNotEmpty();
    }

    @Test
    void ruleIdsAreUnique() {
        Set<String> seen = new HashSet<>();
        for (ConventionRule rule : catalog.all()) {
            assertThat(seen.add(rule.getId()))
                    .as("duplicate rule id: %s", rule.getId())
                    .isTrue();
        }
    }

    @Test
    void everyCheckTypeResolves() {
        for (ConventionRule rule : catalog.all()) {
            assertThat(CheckRegistry.has(rule.getCheckType()))
                    .as("rule %s references unknown checkType '%s'", rule.getId(), rule.getCheckType())
                    .isTrue();
        }
    }

    @Test
    void javadocRuleIsShapedAsDocumented() {
        ConventionRule rule = catalog.byId("JAVADOC_METHOD");
        assertThat(rule).isNotNull();
        assertThat(rule.getCheckType()).isEqualTo("javadocRequired");
        assertThat(rule.isEnabled()).isTrue();
        assertThat(rule.getSeverity()).isEqualTo("WARNING");
        assertThat(rule.appliesToFileType("java")).isTrue();
        assertThat(rule.appliesToFileType("robot")).isFalse();
        assertThat(rule.getMessage()).contains("{{method}}");
        assertThat(rule.getStatus()).isEqualTo("DONE");
    }
}

File: vscode-extension/test/convention-catalog.test.ts
import { describe, it, expect } from 'vitest';
import { loadCanonicalConventions } from './helpers';
import { getCheck, hasCheck } from '../src/convention/check-registry';
import { appliesToFileType } from '../src/convention/convention-rule';

describe('ConventionCatalog (canonical conventions.yaml)', () => {
  const catalog = loadCanonicalConventions();

  it('loads every rule as DONE with a registered check implementation', () => {
    const rules = catalog.all();
    expect(rules.length).toBeGreaterThanOrEqual(14);
    for (const rule of rules) {
      expect(rule.status).toBe('DONE');
      expect(hasCheck(rule.checkType)).toBe(true);
      expect(getCheck(rule.checkType)).toBeDefined();
    }
  });

  it('restricts Robot rules to .robot files and Java rules to .java files', () => {
    for (const rule of catalog.all()) {
      if (rule.checkType.startsWith('robot')) {
        expect(appliesToFileType(rule, 'robot')).toBe(true);
        expect(appliesToFileType(rule, 'java')).toBe(false);
      } else {
        expect(appliesToFileType(rule, 'java')).toBe(true);
        expect(appliesToFileType(rule, 'robot')).toBe(false);
      }
    }
  });

  it('looks up a known rule by id', () => {
    const ids = catalog.all().map((r) => r.id);
    expect(ids.length).toBe(new Set(ids).size);
    const first = catalog.all()[0];
    expect(catalog.byId(first.id)).toEqual(first);
  });
});
